collect/src/main/kotlin/com/certora/collect/TreapListNode.kt

```diff
@@ -0,0 +1,428 @@
+
+    override fun equals(other: Any?): Boolean = when {
+        this === other -> true
+        other is List<*> -> this.size == other.size && this.zip(other).all { it.first == it.second }
+        else -> false
+    }
+
+    override fun isEmpty(): Boolean = false
+
+    override fun builder(): TreapList.Builder<E> = TreapListBuilder(this)
+
+    override fun clear() = EmptyTreapList<E>()
+
+    override fun addFirst(element: E): TreapListNode<E> = TreapListNode(element) append this
+    override fun addLast(element: E): TreapListNode<E> = this append TreapListNode(element)
+
+    override fun addAll(elements: Collection<E>): TreapList<E> = when {
+        elements.isEmpty() -> this
+        else -> this append elements.toTreapList() as TreapListNode<E>
+    }
+
+    override fun add(index: Int, element: E): TreapList<E> = validPosition(index).let {
+        when {
+            index == size -> addLast(element)
+            index == 0 -> addFirst(element)
+            else -> split(index).let { (l, r) -> l!! append TreapListNode(element) append r!! }
+        }
+    }
+
+    override fun addAll(index: Int, c: Collection<E>): TreapList<E> = validPosition(index).let {
+        when {
+            index == size -> addAll(c)
+            c.isEmpty() -> this
+            c is TreapListNode<E> -> when {
+                index == 0 -> c append this
+                else -> split(index).let { (left, right) -> left!! append c append right!! }
+            }
+            else -> split(index).let { (left, right) ->
+                left.orEmpty().addAll(c).addAll(right!!)
+            }
+        }
+    }
+
+    override fun remove(element: E): TreapList<E> = removeNode(element).orEmpty()
+    override fun removeFirst(): TreapList<E> = removeFirstNode().orEmpty()
+    override fun removeLast(): TreapList<E> = removeLastNode().orEmpty()
+    override fun removeAt(index: Int): TreapList<E> = removeNodeAt(validIndex(index)).orEmpty()
+    override fun removeAll(elements: Collection<E>): TreapList<E> = removeAll { it in elements }
+    override fun removeAll(predicate: (E) -> Boolean): TreapList<E> = removeAllNodes(predicate).orEmpty()
+    override fun retainAll(elements: Collection<E>): TreapList<E> = removeAll { it !in elements }
+
+    override fun get(index: Int): E = when (index) {
+        0 -> first()
+        size - 1 -> last()
+        else -> getNodeAt(validIndex(index)).elem
+    }
+
+    override fun set(index: Int, element: E): TreapList<E> = setNodeAt(validIndex(index), element)
+
+    override fun first(): E = firstNode().elem
+    override fun firstOrNull(): E? = first()
+    override fun last(): E = lastNode().elem
+    override fun lastOrNull(): E? = last()
+
+    override fun indexOf(element: E): Int = indexOfFirstNode(rootIndex(), element) ?: -1
+    override fun lastIndexOf(element: E): Int = indexOfLastNode(rootIndex(), element) ?: -1
+
+    override fun contains(element: E): Boolean =
```

Comment:
imho using `forEachElement` here would be nice.

Reply:
Same reasoning about recursion vs. closure allocation here.

Reply:
yah, I see what you mean (though I am sure what the costs are). 
I guess Ideally, we can write a `forEachElement` which is an inline function, but that would mean we need to write it non-recursively - I guess much like the iterator code. 

Of course then it can't be public? 

anyway, probably not worth the effort - though I do find it much more readable.